model/boost_python/spectrum.cc

```diff
@@ -0,0 +1,90 @@
+/*
+ * spectrum.cc
+ */
+#include <boost/python.hpp>
+#include <boost/python/def.hpp>
+#include <string>
+#include <sstream>
+#include <scitbx/constants.h>
+#include <dxtbx/model/spectrum.h>
+#include <dxtbx/model/boost_python/to_from_dict.h>
+#include <scitbx/array_family/boost_python/flex_wrapper.h>
+
+namespace dxtbx { namespace model { namespace boost_python {
+
+  using namespace boost::python;
+  using scitbx::deg_as_rad;
+  using scitbx::rad_as_deg;
+
+  std::string spectrum_to_string(const Spectrum &spectrum) {
+    std::stringstream ss;
+    ss << spectrum;
+    return ss.str();
+  }
+
+  struct SpectrumPickleSuite : boost::python::pickle_suite {
+    static boost::python::tuple getinitargs(const Spectrum &obj) {
+      return boost::python::make_tuple(obj.get_energies(),
+                                       obj.get_weights());
+    }
+
+    static boost::python::tuple getstate(boost::python::object obj) {
+      //const Spectrum &spectrum = boost::python::extract<const Spectrum &>(obj)();
+      return boost::python::make_tuple(obj.attr("__dict__"));
+                                       //beam.get_s0_at_scan_points());
+    }
+
+    static void setstate(boost::python::object obj, boost::python::tuple state) {
+      //Spectrum &spectrum = boost::python::extract<Spectrum &>(obj)();
+      DXTBX_ASSERT(boost::python::len(state) == 2);
+
+      // restore the object's __dict__
+      boost::python::dict d =
+        boost::python::extract<boost::python::dict>(obj.attr("__dict__"))();
+      d.update(state[0]);
+
+      // restore the internal state of the C++ object
+      //scitbx::af::const_ref<vec3<double> > s0_list =
+      //  boost::python::extract<scitbx::af::const_ref<vec3<double> > >(state[1]);
+      //beam.set_s0_at_scan_points(s0_list);
+    }
+
+    static bool getstate_manages_dict() {
+      return true;
+    }
+  };
+
+  template <>
+  boost::python::dict to_dict<Spectrum>(const Spectrum &obj) {
+    boost::python::dict result;
+    result["energies"] = obj.get_energies();
+    result["weights"] = obj.get_weights();
+    return result;
+  }
+
+  template <>
+  Spectrum *from_dict<Spectrum>(boost::python::dict obj) {
+    Spectrum *s = new Spectrum(
+      boost::python::extract<vecd>(obj["energies"]),
```

Comment:
Not convinced about the `vecd` typedef as alias for `af::shared<double>` - we don't _usually_ alias `af::shared` etc. types (though I am receptive to other points of view on this)

I appreciate that it makes the subsequent code a lot shorter 🙂

Reply:
Yah, I grepped cctbx for this and found an instance I liked.  I think the clarity is nice.